pkg/controllers/node/termination/terminator/terminator.go

```diff
@@ -115,18 +131,67 @@ func (t *Terminator) Evict(pods []*v1.Pod) {
 			}
 		}
 	}
-	// 2. Evict in order:
-	// a. non-critical non-daemonsets
-	// b. non-critical daemonsets
-	// c. critical non-daemonsets
-	// d. critical daemonsets
-	if len(nonCriticalNonDaemon) != 0 {
-		t.evictionQueue.Add(nonCriticalNonDaemon...)
-	} else if len(nonCriticalDaemon) != 0 {
-		t.evictionQueue.Add(nonCriticalDaemon...)
-	} else if len(criticalNonDaemon) != 0 {
-		t.evictionQueue.Add(criticalNonDaemon...)
-	} else if len(criticalDaemon) != 0 {
-		t.evictionQueue.Add(criticalDaemon...)
+
+	// EvictInOrder evicts only the first list of pods which is not empty
+	// future Evict calls will catch later lists of pods that were not initially evicted
+	t.EvictInOrder(
+		nonCriticalNonDaemon,
+		nonCriticalDaemon,
+		criticalNonDaemon,
+		criticalDaemon,
+	)
+}
+
+func (t *Terminator) EvictInOrder(pods ...[]*v1.Pod) {
+	for _, podList := range pods {
+		if len(podList) > 0 {
+			// evict the first list of pods that is not empty, ignore the rest
+			t.evictionQueue.Add(podList...)
+			return
+		}
+	}
+}
+
+func (t *Terminator) DeleteExpiringPods(ctx context.Context, pods []*v1.Pod, nodeGracePeriodExpirationTime *time.Time) error {
+	for _, pod := range pods {
+		// check if the node has an expiration time and the pod needs to be deleted
+		deleteTime := t.podDeleteTimeWithGracePeriod(nodeGracePeriodExpirationTime, pod)
+		if deleteTime != nil && time.Now().After(*deleteTime) {
+			t.recorder.Publish(terminatorevents.DeletePod(pod))
+
+			// delete pod proactively to give as much of its terminationGracePeriodSeconds as possible for deletion
+			// ensure that we clamp the maximum pod terminationGracePeriodSeconds to the node's remaining expiration time in the delete command
+			opts := &client.DeleteOptions{
+				GracePeriodSeconds: ptr.Int64(int64(time.Second * time.Until(*nodeGracePeriodExpirationTime))),
+			}
+			if err := t.kubeClient.Delete(ctx, pod, opts); err != nil {
+				if !apierrors.IsNotFound(err) { // ignore 404, not a problem
+					log.FromContext(ctx).WithValues(
+						"namespace", pod.Namespace,
+						"name", pod.Name,
+					).Error(err, "deleting pod")
+				}
+				return err
+			}
```

Comment:
Why does this need to be its own log? Can we just bubble up the error and wrap it? 

Also I believe this is a util function you can use (forgive me if I'm wrong), which should simplify the code
```
			if err := t.kubeClient.Delete(ctx, pod, opts); apierrors.IgnoreNotFound(err) != nil {
```

Reply:
I don't think we want to bubble up the error because it's not blocking, but having a log of the occurrence could catch issues with unexpected cases that users may report on in their logs. I can pull it out if it feels unnecessary though.

Reply:
I'm actually wondering if this could end up being spurious. I'm thinking of a pod which was just successfully evicted but Karpenter's not yet aware. In that case we would go to delete the pod but get a 404. Based on that it might just make sense to drop this.